### ISSUE TITLE
signal.impulse initial condition error

### ISSUE THREAD

**Issue by skulumani** [scipy/scipy#7092]
There seems to be an error in the impulse response function in handling the initial conditions. 

Specifically `impulse` will provide the following error

~~~
ValueError: could not broadcast input array from shape (2,2) into shape (2)
~~~
while `impulse2` works correctly.

Here's an example showing the issue.

~~~
%matplotlib inline
import numpy as np
from scipy import signal
import matplotlib.pyplot as plt

num = 1
den = [1, 4, 5]
time = np.linspace(0,5,100)
sys = signal.TransferFunction(num,den)

t, im2 = signal.impulse(sys,X0=(xd0, x0), T=time)
t, im2 = signal.impulse2(sys,X0=(xd0, x0), T=time)
~~~

Also it seems that the initial conditions are in the wrong order. First, the order of the initial conditions is not specified in the documentation, and it also seems to be the opposite from convention. For example, the matlab function [`intial`](https://www.mathworks.com/help/control/ref/initial.html) uses the following format which seems more intuitive:

~~~
a = [-0.5572, -0.7814; 0.7814, 0];
c = [1.9691  6.4493];
x0 = [1 ; 0];

sys = ss(a,[],c,[]);
initial(sys,x0)
~~~

**Comment by skulumani** [comment]
Also if this is a valid issue, I'd like to take a shot at trying to contribute. I'm fairly new to `python` and `scipy` but would be interested in learning how to help the platform.

**Comment by Eric89GXL** [comment]
@befelix @cwrowley any idea on this one?

@skulumani even if there is no bug, it would be good to clarify the order in the docstring. We can't change the order for backward compatibility reasons.

**Comment by skulumani** [comment]
Thanks for the comment. I can work to update the documentation and submit a
pull request.

**Comment by befelix** [comment]
So looking at the code the logic that handles `X0` is a little bit different in both cases and I think it could probably be made more robust and compatible without affecting backwards compatibility (see [impulse](https://github.com/scipy/scipy/blob/98e2dcf8c083bb94542b6dddbc89d377556dc472/scipy/signal/ltisys.py#L2115) vs [impulse2](https://github.com/scipy/scipy/blob/98e2dcf8c083bb94542b6dddbc89d377556dc472/scipy/signal/ltisys.py#L2201)).

The order of initial conditions is mostly dictated by `tf2ss`, which is used under the hood to convert to a state space system that can be simulated using `lsim`/`lsim2`.

**Comment by cwrowley** [comment]
Yes, it looks like as currently implemented, `X0` needs to be given as a column vector.  It would probably be good to allow this to be either a column vector or a 1d array.  For instance, as currently implemented, the following works (with `X0` as a column vector):

```
t, y = signal.impulse(sys, X0=np.array([[0];[0]]))
```
but the following do not (with X0 as a row vector or 1d array):
```
t, y = signal.impulse(sys, X0=np.array([[0],[0]]))
t, y = signal.impulse(sys, X0=(0,0))
```
Also, note that the error occurs within [lsim](https://github.com/scipy/scipy/blob/98e2dcf8c083bb94542b6dddbc89d377556dc472/scipy/signal/ltisys.py#L1948), and should be easy to fix (just reshaping `X0` within `lsim`).

**Comment by skulumani** [comment]
Following on from @cwrowley here are some of the differences in the format of the initial conditons for `impulse` and `impulse2` that I could think of

~~~
# _, tf_im = signal.impulse(systf, X0=(xd0, x0), T=time) # doesn't work
# _, tf_im = signal.impulse(systf, X0=np.array([xd0, x0]), T=time) # doesn't work
_, tf_im = signal.impulse(systf, X0=np.array([[xd0],[x0]]), T=time) 

_, tf_im2 = signal.impulse2(systf, X0=(xd0, x0), T=time) 
# _, tf_im2 = signal.impulse2(systf, X0=np.array([xd0, x0]), T=time) 
# _, tf_im2 = signal.impulse2(systf, X0=np.array([[xd0],[x0]]), T=time) # doesn't work
~~~

I think there's another issue #5676 which seeks to address some of this.

@befelix I think I found the differences in the initial state convention. I'm used to dealing with this form of the [controllable canonical realization](https://en.wikipedia.org/wiki/State-space_representation#Canonical_realizations). Specifically, the characteristic equation becomes the bottom row of the `A` matrix. This has been the form I have most often seen but on the other hand, both `scipy` and `matlab` default to a different companion realization which reverses the order of the states. As a result, the order of the initial states are also reversed, ie. highest order comes first.

The issue was mostly due to myself not checking the form of the realization. At the very least some updated documentation might be helpful. For example, we could show an example transfer function and it's corresponding realization in state space form. In addition, the order of the initial conditions could be documented in all of the appropriate functions.

Finally, what's the best way to clone `scipy` and work on a local copy without messing up the version already installed?

**Comment by befelix** [comment]
Great, please reference this issue when you open a PR. 

Regarding development have a look at https://www.scipy.org/scipylib/building/ - if you don't want to mess up your local version have a look at `virtualenv` or the equivalent in anaconda.
